topics/engineering/running-background-processes-in-python.gmi

```diff
@@ -0,0 +1,68 @@
+# Running Background Processes in Python using Popen
+
+* author: bonfacem
+* reviewed-by: jnduli
+
+Recently, I needed to write test fixtures in Python to run RDF tests against a Virtuoso[0] instance running in the background.  The correct command to run virtuoso in the background is:
+
+```
+# Replace <init-file> with the correct path to your virtuoso ini file
+virtuoso-t +wait +no-checkpoint +configfile <init-file>
+```
+
+The complete snippet for the aforementioned fixture is outlined below:
+
+```
+ 1         # Here we intentionally ignore the "+foreground" option to
+ 2         # allow virtuoso to run in the background.
+ 3         with subprocess.Popen(
+ 4             [
+ 5                 "virtuoso-t",
+ 6                 "+wait",
+ 7                 "+no-checkpoint",
+ 8                 "+configfile",
+ 9                 init_file,
+10             ],
+11             stdout=subprocess.DEVNULL,
+12             stderr=subprocess.DEVNULL,
+13         ) as pid:
+14             pid.wait()
+15 	    [...]
+16             yield response
+17             [...]
+18             for pid_ in get_process_id(init_file):
+19                 psutil.Process(pid_).kill()
+```
+
+Using "subprocess.Popen"(L3-L13), we spawn the Virtuoso process in the background.  However, after the tests finished, the Virtuoso process remained running in the background, leading to failures in subsequent tests.  To handle this, we had to manually kill it(L18-L19).  This approach is clanky because we expected the Virtuoso process to terminate when the parent process ended.  Calling "pid.terminate()" didn’t work since Virtuoso was spawned under a different PID without a parent.
```

Comment:
I don't think the virtuoso process starts in the background, but rather is daemonized. This means once the tests finish, the virtuoso process can't be killed because it isn't a child of pytest.

Also, it isn't that virtuoso was spawned under a different PID. Here's a video explaining the process: https://youtu.be/hgb-1f6BlGI?list=PLSIUOFhnxEiC3YTdxwqZqgEY5imVL8U8J&t=1120